internal/privatestate/data.go

```diff
@@ -0,0 +1,163 @@
+package privatestate
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"strings"
+	"unicode/utf8"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+)
+
+// Data contains private state data for the framework and providers.
+type Data struct {
+	// Potential future usage:
+	// Framework contains private state data for framework usage.
+	Framework map[string][]byte
+
+	// Provider contains private state data for provider usage.
+	Provider ProviderData
+}
+
+// Bytes returns a JSON encoded slice of bytes containing the merged
+// framework and provider private state data.
+func (d Data) Bytes(_ context.Context) ([]byte, diag.Diagnostics) {
+	var diags diag.Diagnostics
+
+	bytes, err := json.Marshal(d)
+	if err != nil {
+		diags.AddError(
+			"Error Encoding Private State",
+			fmt.Sprintf("An error was encountered when encoding private state: %s.\n\n"+
+				"This is always a problem with Terraform or terraform-plugin-framework. Please report this to the provider developer.", err),
+		)
+
+		return nil, diags
+	}
+
+	return bytes, diags
+}
+
+// NewData creates a new Data based on the given slice of bytes.
+// It must be a JSON encoded slice of bytes, that is map[string][]byte.
+func NewData(ctx context.Context, data []byte) (Data, diag.Diagnostics) {
+	var (
+		u     map[string][]byte
```

Comment:
Nit: On first read I wasn't sure what `u` was here -- should we call it something like "dataMap"? 👍